### DISCUSSION TITLE
Configurable repository cache root directory path

### DISCUSSION THREAD

**Discussion by Gabriel-Ladzaretti** [renovatebot/renovate#32444] [Suggest an Idea]
### Tell us more.

It is currently impossible to bind mount a repo cache file at the expected path when running a Renovate job as a docker container. this is because, if parent directories are not present in the container, they are created with root as the owner by default, and this docker behavior is not configurable. one such directory is the main cache directory, and when the cli tries to create a subdir, it fails with a permission error (e.g., `__private_cache_dir`).

one approach to address this limitation is to make the repo cache root directory configurable. 
a current workaround is to bind mount the cache directory as well as the repo cache file, but this requires cleanup on the host, which is sub-optimal.

**Comment by rarkins** [comment]
What is the current approach/default Renovate uses?

**Reply by Gabriel-Ladzaretti** [reply]
currently the repo cache is created and managed by the cli.
and saved under `<CACHE_DIR>/renovate/repository/<PLATFORM>/<ORG>/<REPO>.json`

**Reply by rarkins** [reply]
And are you trying to bind a mount at `<CACHE_DIR>/renovate/repository/<PLATFORM>/<ORG>/`?

**Reply by rarkins** [reply]
At which level of this path are you wanting to make it newly configurable?

**Reply by Gabriel-Ladzaretti** [reply]
> And are you trying to bind a mount at `<CACHE_DIR>/renovate/repository/<PLATFORM>/<ORG>/`?

yes, we bind the actual cache file at this path.
no other way to bind the repo cache, as this is the path it is looked for.

> At which level of this path are you wanting to make it newly configurable?

parent. as it seems the easiest and the most straight forward.

i,e,.
 `<REPO_CACHE_DIR>/renovate/repository/<PLATFORM>/<ORG>/<REPO>.json`

**Reply by rarkins** [reply]
Or `<REPO_CACHE_DIR>/<PLATFORM>/<ORG>/<REPO>.json` ?

**Reply by Gabriel-Ladzaretti** [reply]
i dont mind either, but i suppose it depends on how much we want to change this function

https://github.com/renovatebot/renovate/blob/97ace64c0a249abfa7f78be9bed1bca21b5833bb/lib/util/cache/repository/common.ts#L6-L13

**Reply by rarkins** [reply]
I think we can refactor that fairly safely to then make repoCacheDir make more sense

**Reply by Gabriel-Ladzaretti** [reply]
what should be the default value for `repoCacheDir`?

**Reply by rarkins** [reply]
I just realized that this logic is nested pretty deep. At the lowest level it's calling https://github.com/renovatebot/renovate/blob/3a516ff10de2540ac33d6741e2ec66ed8413b18a/lib/util/fs/util.ts#L16-L23 which assumes it's inside `baseDir`. But presumably you don't want it inside `baseDir`?

**Reply by Gabriel-Ladzaretti** [reply]
it depends really, if `baseDir` already exists in the container, it wont have a problem being its child.
otherwise, it will fail the same way it does currently.
so i guess the best approach would be to use a different dir to avoid such cases.
note that this problem is specific to the docker executor only, if that matters.

**Reply by Gabriel-Ladzaretti** [reply]
is a dedicated config opt that defaults to the cache dir an acceptable solution ?

**Reply by rarkins** [reply]
Either way is ok

**Reply by Gabriel-Ladzaretti** [reply]
Actually, I much prefer a "no code" solution. as you said, most of the current logic resolving directories is pretty low level.

one such solution is to have the `baseDir` and the `<baseDir>/cache` directories pre-exist inside the container with `0775` permissions and `:root` ownership. this way, we won't have issues with write permissions. doing so enables us to both bind mount the repo cache file from outside the container and provide the appropriate directory permissions for renovate.

above means that the only change needed is creating these directories in the renovate docker image.

wdyt ?

**Reply by rarkins** [reply]
@viceice in which repo would this best be done?

**Reply by viceice** [reply]
i think it's best suited in renovate images.
